### DISCUSSION TITLE
oss-fuzz

### DISCUSSION THREAD

**Discussion by m-holger** [qpdf/qpdf-dev#6] [General]
Current state of play as far as I can make out is: There are 7 open cases. None can be reproduced reliably any longer. None have crashed since 2 July and hopefully all will be auto-deleted at various dates between 7 July and 16 July.

**Comment by jberkenbilt** [comment]
That would be excellent. There is a note in TODO.md about some tests that are being ignored. That comment can be removed now. Thanks for tackling this.

**Comment by m-holger** [comment]
4 of the 7 open cases have popped up again. They showed no crashes since 2 July when I checked yesterday morning, but now they show crashes on various dates that previously were reported as clean. The only explanation I can come up with is that there are sometimes long and highly variable delays in updating the crash stats. The remaining 3 cases have unchanged auto-delete dates.

One of the 4 case is 70049. I merged https://github.com/qpdf/qpdf/pull/1230 before this mornings build  which seems to have fixed this case. I will keep an eye on the remaining three. 

I am looking at 70055.

**Reply by m-holger** [reply]
70055 is a xref stream with a section with a negative number of entries. I will push a PR this evening my time and merge before tomorrows build.

**Reply by jberkenbilt** [reply]
Thanks! Clearly both fuzzing FUTURE and adding you as a maintainer were good choices. :-)

**Reply by m-holger** [reply]
> 4 of the 7 open cases have popped up again. 

Duh!  The clue lies in "I merged https://github.com/qpdf/qpdf/pull/1230 before this mornings build"

**Comment by m-holger** [comment]
Update

https://github.com/qpdf/qpdf/pull/1230 omitted a limit on the size of uncompressed JPEG images. This was added back in https://github.com/qpdf/qpdf/pull/1234. The omission is likely to have contributed to previously fixed cases popping up again. (However, this can not explain why prior periods that did not show crashes suddenly showed crashes.)

A change introduced in https://github.com/qpdf/qpdf/pull/1230 was that damaged JPEGs throw runtime errors. This was implemented by checking for warnings after each call to jpeglib procedures. It subsequently transpired that some procedures (particularly jpeg_read_header) generate thousands of warnings and cause a timeout. To deal with this,  https://github.com/qpdf/qpdf/pull/1236 provided a custom implementations of jpeglib's emit_message that throws a runtime error as soon as the first warning is encountered. This dealt with some of the cases. There were others however where a significant amount of time elapsed before the first warning was emitted. To attempt to deal with this, https://github.com/qpdf/qpdf/pull/1239 reduced the max size of uncompressed JPEGs during fuzzing to 50MB.

There are other reasons for time-outs, including

- large highly damaged files resulting in cases timing out in xref reconstruction. To deal with this, https://github.com/qpdf/qpdf/pull/1239 added a sanity check during xref reconstruction, giving up after 1000 warnings are generated. To me it makes sense to generally give up if such a high number of warnings is generated, however the limit could be restricted to fuzzing only.
- damaged files with non-dictionaries in pages tree.  https://github.com/qpdf/qpdf/pull/1237 adds code to exclude non-dictionaries when building the all_pages vector. This should probably be extended to actually remove them from the pages tree in order to avoid qpdf writing invalid pages trees. https://github.com/qpdf/qpdf/pull/1241 adds calling getAllPages to the xref reconstruction sanity checks.
- recovery from parsing errors. At the moment, the parser has no overall error limit, and qpdf accepts objects that may have generated 100s of parser errors and may not been followed bu 'endobj'.
- createFromJSON and updateFromJSON continue parsing if errors are detected, even though the ultimate result is that an exception is thrown. It would make sense to throw the exception immediately. If this is not acceptable, a warning limit during fuzzing is needed to avoid time-outs.

**Comment by jberkenbilt** [comment]
This is a good summary. I am still uncomfortable with using a separate repo for this. I'll comment on your other issue.

**Comment by m-holger** [comment]
Update

Adding to the removal of non-dictionary objects from the all_pages vector, https://github.com/qpdf/qpdf/pull/1242 adds code to (if necessary) call flattenPagesTree to remove invalid objects from the pages tree. The PR also limits the number of warnings produced by the json_fuzzer before throwing an exception. (Throwing an immediate exception when an error is detected would not be desirable for fuzzing as it would make it more likely for trivial input errors to mask more interesting errors. It is not possible to throw immediately during CI as many test encounter other errors before encountering the error tested for.)

https://github.com/qpdf/qpdf/pull/https://github.com/qpdf/qpdf/pull/1247 adds hard limit on the number of warnings encountered by qpdf_fuzzer in order to avoid time outs based on the sheer number of invalid objects encountered. Earlier attempts to use a hard limit during xref recovery has proved insufficient. It also removes the soft limit of only displaying a limited number of warnings during fuzzing. While the sheer number of warnings generated are usually counterproductive for analysing problems, unfortunately sometimes oss-fuzz produces  empty test cases. In those cases the warnings become the only clue of what the test case might have contained. 

The PR  also limits the memory used by Pl_PNGFilter and Pl_Tiffpredictor in the same way Pl_DCT. There is an indication that in one case a (or probably two) 0.5GB buffer was allocated during Flate decoding, presumably because of an invalid /Columns entry. (Unable to confirm as test case is empty.) 

The current state is: there are no cases that can be reliably reproduced. The last dct_fuzzer timeout was last Friday morning. The last json_fuzzer timeout was before limiting the number of warnings. Issues with the qpdf_fuzzer are continuing.

There are a number of difficulties tracking down what the issues are, The most significant problem is that is impossible to identify which test case is timing out / running out of memory. For example, all unreproducible timeouts of the qpdf_fuzzer are shown as issue 17630 with the test case produced in 2019. This case was a timeout as the result of a corrupt JPEG and most probably has been fixed. There was another new reproducible case on Tuesday that was initially shown separately, but also was shown as a crash for 17630. Later on Tuesday the case was linked to 17630, and both cases where visible for 17630, one as reproducible and the 2019 case as flaky. The new case was fixed by  https://github.com/qpdf/qpdf/pull/1242 . oss-fuzz showed the case as fixed for a while, and now the test case has completely disappeared from all listings (open, closed or all cases). The crash is still shown as reproducible but links back only to the 2019 test case (which is correctly shown as flaky) which has nothing to do with the crash other than that it happened to the same fuzzer and was also a timeout. There are other examples of the same happening

I don't think that oss-fuzz list any additional flaky cases while the existing flaky case i open, but keeps the existing flaky case open because of those other crashes. This makes it both difficult to resolve the case and important because the underlying reasons of those other crashes may be different and more important.

**Comment by m-holger** [comment]
Update

There were no crashes on yesterday's build (i.e. including  https://github.com/qpdf/qpdf/pull/1242). I am mildly optimistic, although for some of the cases crashes were very intermittent, and even the regular cases may be a mixture of frequent and more intermittent crashes. A much longer run is therefore needed to have some degree of certainty.

In any case, its probably time to switch to dealing with future rather than current issues. Key areas that I think may cause future issues are

- invalid trailers
- invalid xref tables
- invalid [page trees](https://github.com/qpdf/qpdf-dev/discussions/10)
- invalid filtered streams
- excessive error recovery in parser and tokenizer
- invalid JSON input

Luckily this is quite a short list.

**Comment by jberkenbilt** [comment]
Thanks. No objections to any of this.

**Comment by m-holger** [comment]
The bad news: there are still crashes (although greatly reduced).

The good news: I finally spotted where I can find the test cases that actually crashed rather than a test case that 5 years ago crashed.

Current crashes are due to a ridiculously large filter /Column value as suspected and a false positive in MD5 calculations.

**Reply by jberkenbilt** [reply]
It's like with plumbing. The water seldom drips from a spot right near the leak.

**Reply by jberkenbilt** [reply]
P.S. I'm very grateful for your digging into this.

**Comment by m-holger** [comment]
> However, this can not explain why prior periods that did not show crashes suddenly showed crashes

Mystery solved: if you keep the "Statistics" table open in a browser tab to monitor crashes, the "END TIME" in the table heading gets updated every hour, but not the table itself. A clean table will stay clean indefinitely.

**Comment by m-holger** [comment]
There were two crashes in the last 72 hours:

- dct_fuzzer time-out. The test case for the crash does not seem to match up with the log, as the test case throws an error before the point where the test allegedly hangs (and the log shows that the same error occurred multiple times before without hanging). As this is the only dct_fuzzer timeout in the last 14 days I am going to ignore this for now.

- qpdf_fuzzer out of memory. This was caused by a dangling reference with a very high object id and will be solved  by  https://github.com/qpdf/qpdf/pull/1171 (needs review).

**Comment by m-holger** [comment]
There were some further crashes, including:

- qpdf_fuzzer time-out due to a very slow  running reconstruct_xref. https://github.com/qpdf/qpdf/pull/1255 includes some refactoring which for the pathological test case reduces runtime on my system from > 70secs to < 0.03 secs.

- qpdf_fuzzer out of memory due to a corrupt Flate filtered stream. I have added a memory limit for Pl_Flate to qpdf_fuzzer.

**Comment by m-holger** [comment]
There were some further crashes, including:

- qpdf_fuzzer time_out due to slow Pl_RunLength::decode. https://github.com/qpdf/qpdf/pull/1258 includes local buffering of the output which significantly improves the runtime for the test case.

- there was another dct_fuzzer as well as qpdf_fuzzer time-out due to corrupt jpegs. (dct_fuzzer had two time-outs in the last 14 days). https://github.com/qpdf/qpdf/pull/1258 reduces the memory limits during fuzzing.

- there were a couple of time-outs due to large corrupt flate filtered streams, including content streams. The time-outs were partly caused due to QPDFWriter repeatedly testing whether the streams can be filtered (addressed in https://github.com/qpdf/qpdf/pull/1257) and partly due to the time it takes to normalize the content streams (addressed by adding a maximum size for flate filtered streams during fuzzing in https://github.com/qpdf/qpdf/pull/1257 and https://github.com/qpdf/qpdf/pull/1258).

**Comment by m-holger** [comment]
There was another flaky dct_fuzzer out-of-memory  crash with an empty test input. The log file for the case does not make much sense in sofar as it indicates that the allocations were in jpeglib which is not called with empty inputs. However, looking at the code that is actually execute it transpired that the behaviour of std::string::append with a zero length buffer is undefined, although it did behave as expected. I have added guards against this case in https://github.com/qpdf/qpdf/pull/1263 .

**Comment by m-holger** [comment]
There was another qpdf_fuzzer time-out due to a corrupt flate filtered stream. Processing was actually aborted due to the memory limit set in the fuzzer, but the subsequent call to finish was slow, taking up around 90% of the total runtime.  https://github.com/qpdf/qpdf/pull/1265 adds a check to the finish method to prevent this.

This raises the wider question: is there a reason to attempt to call the pipeline finish method in `QPDF::::pipeStreamData` if filtering is aborted due to an exception?

**Comment by m-holger** [comment]
There was a single crash in the last week. qpdf_fuzzer timed out in jpeg_start_decompress. There was nothing either in the size of the input or in the cinfo returned by jpeg_read_header to indicate that there would be a problem. I don't think it is worth (both in terms of effort and in terms of added complexity) trying to prevent such time-outs given the frequency of occurrence.

**Reply by jberkenbilt** [reply]
I had reached a similar conclusion about the issues that had been open prior to your spending time on this, so I have no issues with leaving these alone. I think you made meaningful progress, but ultimately, it's not our job to fuzz the jpeg library, and there are legitimate cases where qpdf may take a long time.

**Reply by m-holger** [reply]
>  ultimately, it's not our job to fuzz the jpeg library

I fully agree. qpdf probably spends too much time fuzzing other libraries rather than its own code. The tricky bit is to limit the time without preventing the fuzzer from finding problems in how qpdf calls the libraries.

One of the reasons why qpdf spends a disproportionate amount of time on libraries is that there is a single qpdf fuzzer plus a json fuzzer vs 6 filter fuzzers, giving the core qpdf fuzzer 1/8 of the total time. Splitting the qpdf fuzzer into 6 separate fuzzers for each of the component jobs would help to redress the balance. 

>  there are legitimate cases where qpdf may take a long time.

Agreed. Splitting the qpdf fuzzer would help here too, as in many legitimate cases the individual jobs would not time out.

**Reply by m-holger** [reply]
The time-outs in jpeg_read_header were due to corrupt progressive jpegs with thousands of passes (instead of the more normal 3 - 5). https://github.com/qpdf/qpdf/pull/1275 traps the issue. This is probably something that could be enabled for qpdf generally instead of just for the fuzzers.

**Comment by m-holger** [comment]
There were another two qpdf_fuzzer time-outs 5 days ago.

The first time-out was due to a problem in with QPDF::reconstruct_xref, where  [this call to read_xref](https://github.com/qpdf/qpdf/blob/0b3debaf86eda5ecd1dd6447bcf2ac62abb6dd16/libqpdf/QPDF.cc#L634) attempted to read a corrupt xref stream, which introduced non-existing uncompressed objects into the xref table. https://github.com/qpdf/qpdf/pull/1273 roughly halves the time required to deal with those objects, but the proper long-term fix would be to strengthen the checks during recovery.

The second one was a heavily damaged file with no particular issues. The fuzzer handled the file fine five times, but on the sixth sub-task  it ran out of time at a place that it handled before during the earlier sub-tasks. A similar case popped  up a couple of days ago.

Unless you have any objections, I will split the qpdf_fuzzer in order to avoid this issue.

**Comment by m-holger** [comment]
There are currently no open issues.

The qpdf_fuzzer has been split into 6 seperate fuzzers in https://github.com/qpdf/qpdf/pull/1279 . The PR omitted the code to install the `.options` files for the new fuzzers, which will need to be fixed in due course. 

Following the split, a number of flakey time out/out of memory crashes were seen, particularly in areas that would be reached late in the original combined fuzzer. These cases look different  from the cases encountered before.  I am not sure whether this is because with the combined fuzzer time-outs would have occurred much earlier, or whether it is because the new fuzzers are starting fresh without the large corpus of the original fuzzer. Whatever the reason, the cases highlighted that there is generally a high overhead with processing corrupt object and content streams. To prevent this overhead from masking other issues I reduced the maximum size of Flate filtered streams during fuzzing in https://github.com/qpdf/qpdf/pull/1279 as well as adding a limit on the total number of errors in a single object during parsing in  https://github.com/qpdf/qpdf/pull/1288. (This limit is general rather than fuzz specific. It needs to be decided whether this is appropriate, and if so, whether it should be documented in the change log/release notes.)

Following the split, the fuzzers found two bugs introduced during earlier fuzz related fixes:

- the code to eliminated invalid pages from the pages tree introduced in https://github.com/qpdf/qpdf/pull/1242 did not work correctly if getAllPages threw an exception due to a loop in the pages tree.

- the memory limit on Flate filtered streams causes errors / memory leakage during deflaltion.

Both of these issues were fixed in  https://github.com/qpdf/qpdf/pull/1285.

The fuzzers also found two issues in https://github.com/qpdf/qpdf/pull/1272 , which were fixed in  https://github.com/qpdf/qpdf/pull/1289 , as well as in the code to find the `EI` operator, which was fixed in  https://github.com/qpdf/qpdf/pull/1291 .

An underlying problem highlighted by the new cases  is that in many cases there is no effective limit to length of object that both the tokenizer and parser are going to process, with serious implication for processing times and memory requirements. In the case fixed by  https://github.com/qpdf/qpdf/pull/1291, we had a corrupt content stream rich in 'EI' sequences and short of delimiters - essentially for each 'EI' encountered the tokenizer would read a long sequence of characters. Similarly in an object stream, if there is a missing container close token, the parser will read to the end of the stream, possibly creating thousands of objects in the process. In addition to the performance implications, this may also create undetectable errors as highlighted in   https://github.com/qpdf/qpdf/pull/1266 .

**Comment by jberkenbilt** [comment]
Outstanding progress.

**Comment by m-holger** [comment]
Current state of play is that there are no open tickets. There are also no open cases older than 7 days. There have been 2 flaky timeouts in the last 7 days, with two distinct cases. The first cases will be fixed by the planned additional /Page object sanity checks outlined in #19. I have not fully investigated the second case yet, but it initially points to the need for stronger validations on xref streams and xref tables.

The cases found by oss-fuzz over the last for weeks have all been issues that have been in qpdf for a long time. There are a number of reasons why they are being found now:

- Timeouts/ out-of-memory issues due to corrupt compression have now been largely suppressed. In the past these dominated the cases and any issues due to qpdf causes were lumped in with these and not separately reported and therefore were invisible.

- The split of the qpdf fuzzers has resulted in a higher proportion of time being spent on fuzzing qpdf code vs. external libraries.

- Improvements in recovery means that damaged files exercise more code rather than  failing very early.

**Comment by jberkenbilt** [comment]
Fantastic progress.

**Comment by m-holger** [comment]
We have no open oss-fuzz cases at the moment. While I am sure that this is only temporary, might as well enjoy it while it lasts.

**Reply by jberkenbilt** [reply]
Indeed -- each time we reach this space, we free the fuzzer up to crawl into new corners and find even more ways for us to tighten things up. Definitely worth savoring the silence though.

**Reply by m-holger** [reply]
Well, we knew it wasn't going to last. See #23

**Reply by m-holger** [reply]
Issue  433311400 very much looks like it is #23 (3rd item). I'll fix it Thursday or Friday.
